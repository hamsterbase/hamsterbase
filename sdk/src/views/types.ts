export enum WebpagesViewSortOrder {
  random = 0,
  title_asc = 1,
  title_desc = 2,
  add_time_asc = 3,
  add_time_desc = 4,
  annotate_count_asc = 5,
  annotate_count_desc = 6,
}

export enum WebpagesViewFilterType {
  constant = 0,
  select = 1,
  multiple_selection = 2,
}

export enum AnnotationFilterOperator {
  is = 0,
  is_not = 1,
  is_greater_than = 2,
  is_less_than = 3,
}

export enum FirstAddTimeFilterOperator {
  before = 0,
  after = 1,
}

export enum FirstAddTimeFilterRightOperands {
  day_1 = 0,
  week_1 = 1,
  month_1 = 2,
  day_2 = 3,
  week_2 = 4,
  month_2 = 5,
}

export type WebpagesViewFilter =
  | {
      type: WebpagesViewFilterType.multiple_selection;
      key: 'host';
      operator: boolean;
      rightOperand: string[];
    }
  | {
      type: WebpagesViewFilterType.constant;
      key: 'liked';
      operator: boolean;
    }
  | {
      type: WebpagesViewFilterType.select;
      key: 'annotation';
      operator: AnnotationFilterOperator.is;
      rightOperand: number;
    }
  | {
      type: WebpagesViewFilterType.select;
      key: 'first-add-time';
      operator: FirstAddTimeFilterOperator.before;
      rightOperand: FirstAddTimeFilterRightOperands;
    }
  | {
      type: WebpagesViewFilterType.constant;
      key: 'link-duplicate';
      operator: boolean;
    }
  | {
      type: WebpagesViewFilterType.constant;
      key: 'title-duplicate';
      operator: boolean;
    };

export type WebpagesViewFilterMatchRule = 'and' | 'or';

export interface CreateViewRequest {
  title: string;
  description?: string;
  sort?: WebpagesViewSortOrder;
  limit?: number;
  enableLimit?: boolean;
  filters?: WebpagesViewFilter[];
  match?: WebpagesViewFilterMatchRule;
}

export interface PatchWebpagesViewRequest {
  title?: string;
  description?: string;
  sort?: WebpagesViewSortOrder;
  limit?: number;
  enableLimit?: boolean;
  match?: WebpagesViewFilterMatchRule;
}

export type WithId<T> = T & { id: string };

export interface WebpagesView {
  id: string;
  title: string;
  createdAt: number;
  enableLimit: boolean;
  limit: number;
  sort: WebpagesViewSortOrder;
  filters?: WithId<WebpagesViewFilter>[];
}
