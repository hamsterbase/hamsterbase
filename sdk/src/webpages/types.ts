export interface UploadWebpageRequest {
  title?: string;
  link?: string;
  content: string;
  ext: WebsiteExt;
}

export enum WebsiteExt {
  mhtml = 'mhtml',
  webarchive = 'webarchive',
  html = 'html',
}

export interface WebpageHighlight {
  highlightId: string;
  type: 1;
  marker: {
    meta: {
      textBefore: string;
      textAfter: string;
    };
  };
  text: string;
  note?: string;
  editTime?: number;
}

export interface Webpage {
  title: string;
  excerpt: string;
  firstAddTime: number;
  createTime: number;
  liked: boolean;
  link?: string;
  ext: WebsiteExt;
  id: string;
  size: number;
  read: boolean;
  highlights: WebpageHighlight[];
}

export interface PatchWebPageRequest {
  title?: string;
  excerpt?: string;
  liked?: boolean;
  link?: string;
  read?: boolean;
  labels?: string[];
}

export interface CreateHighlightRequest {
  text: string;
  textBefore: string;
  textAfter: string;
  tenThousandth?: number;
  note?: string;
}

export interface UploadWebpageResponse {
  id: string;
}

export interface WebpagesFilterOptions {
  liked?: boolean;
  read?: boolean;
  annotated?: boolean;
  host?: string | string[];
  labels?: string | string[];
  ext?: WebsiteExt | WebsiteExt[];
  /**
   * @default 30
   */
  per_page?: number;
  /**
   * @default 1
   */
  page?: number;
}

export interface WebpagesListOptions extends WebpagesFilterOptions {
  /**
   * @default 'first_add_time_desc'
   */
  sort?:
    | 'title_asc'
    | 'title_desc'
    | 'first_add_time_asc'
    | 'first_add_time_desc'
    | 'annotate_count_asc'
    | 'annotate_count_desc';
}

export interface WebpagesSearchOptions extends WebpagesFilterOptions {
  q: string;
}

export interface UpdateWebpageHighlightOption {
  note: string;
}

export interface WebpagesList {
  webpages: Webpage[];
}
