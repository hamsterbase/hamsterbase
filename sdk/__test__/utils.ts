import fs from 'fs/promises';
import path from 'path';

export const enum Fixtures {
  HamsterBaseDocument_01_mht = '001.mht',
  HamsterBaseGithubIssue_02_webarchive = '002.webarchive',
  HamsterBaseGithubHome_03_html = '003.html',
}

export const enum FixturesId {
  HamsterBaseDocument_01_mht = 'bcf1e35729685a87ce18733080eaf0f80fec0c81a5a4608ef5b3f0272a37851f',
  HamsterBaseGithubIssue_02_webarchive = 'b03e82c2b622df1f039ecba4910f464becf5600b12425ea3d1f65d4c17d70d8d',
  HamsterBaseGithubHome_03_html = 'd4a1267d1f8a8eeb20d7b4b86a9d91839809e083ceb42366847ed5ed3a6cf254',
}

export function getBase64Fixture(fixtureName: Fixtures): Promise<string> {
  return fs.readFile(path.join(__dirname, 'fixtures', fixtureName), 'base64');
}

export function resolveFixture(fixtureName: string) {
  return path.resolve(__dirname, 'fixtures', fixtureName);
}

export function resolveRoot(relativePath: string) {
  return path.resolve(__dirname, '../..', relativePath);
}

export function resolveBetaFile(relativePath: string) {
  return path.resolve(resolveRoot('hamsterbase-0.5.0-beta'), relativePath);
}

export function getPort() {
  return Math.floor(Math.random() * 4000) + 5000;
}
