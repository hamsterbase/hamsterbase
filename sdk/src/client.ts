import { ErrorCodes, HttpError, ResponseError } from './error';

export interface ClientOptions {
  endpoint: string;
  token: string;
  requestLib?: RequestLib;
}

export interface RequestLib {
  (url: string, options: RequestInit): Promise<Response>;
}

export type HamsterBaseResponse<T> =
  | { success: true; data: T }
  | { success: false; message: string; error: ErrorCodes };

export class Client {
  constructor(private options: ClientOptions) {
    if (!options.endpoint) {
      throw new Error('endpoint is required');
    }
    if (!options.token) {
      throw new Error('token is required');
    }
  }

  post<T>(api: string, data: unknown) {
    return this.request<T>('POST', api, data);
  }

  patch<T>(api: string, data: unknown) {
    return this.request<T>('PATCH', api, data);
  }

  put<T>(api: string, data: unknown) {
    return this.request<T>('PUT', api, data);
  }

  delete<T>(api: string) {
    return this.request<T>('DELETE', api);
  }

  get<T>(api: string) {
    return this.request<T>('GET', api);
  }

  private async request<T>(
    method: string,
    api: string,
    data?: unknown
  ): Promise<T> {
    const requestLib = this.options.requestLib ?? fetch;
    const response = await requestLib(
      `${this.options.endpoint}/public/api/v1${api}`,
      {
        method: method,
        headers: {
          'Content-Type': 'application/json',
          authorization: `Bearer ${this.options.token}`,
        },
        body: data ? JSON.stringify(data) : undefined,
      }
    );
    if (response.status !== 200) {
      const message = await response.text();
      try {
        throw new HttpError(response.status, JSON.parse(message).message);
      } catch (error) {
        if (error instanceof HttpError) {
          throw error;
        }
        throw new HttpError(response.status, message);
      }
    }
    const json = (await response.json()) as HamsterBaseResponse<T>;
    if (!json.success) {
      throw new ResponseError(json.error, json.message);
    }
    return json.data;
  }
}
