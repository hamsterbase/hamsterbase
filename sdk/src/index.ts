export * from './hamsterbase';
export * from './client';
export * from './webpages';
export * from './views';
export * from './error';
