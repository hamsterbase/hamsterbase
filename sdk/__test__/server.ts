import { join } from 'path';
import { fork } from 'child_process';
import { resolveBetaFile, resolveFixture, resolveRoot } from './utils';
import fs from 'fs/promises';

interface RunServerOptions {
  port: number;
  database: string;
}

async function initToken(database: string, token: string) {
  await fs.mkdir(join(database, 'config'), { recursive: true });
  const tokenPath = join(database, 'config/token.json');
  await fs.writeFile(
    tokenPath,
    JSON.stringify({
      token: [
        {
          label: 'test',
          token: token,
          createdAt: Date.now(),
        },
      ],
    })
  );
}

export async function initCache(database: string) {
  await fs.mkdir(join(database, 'cache', 'webpage-parser'), { recursive: true });
  await fs.cp(resolveFixture('webpage-parser'), join(database, 'cache', 'webpage-parser'), { recursive: true });
}

export async function createTestServer(options: RunServerOptions) {
  const token = `${Math.random()}`;
  try {
    await fs.rm(options.database, { recursive: true });
  } catch (error) {}
  await initToken(options.database, token);
  await initCache(options.database);
  return new Promise<{
    endpoint: string;
    token: string;
    dispose: () => Promise<void>;
  }>((resolve) => {
    const mainServer = process.env.HAMSTER_BASE_SERVER || resolveBetaFile('server/main.js');
    const cp = fork(mainServer, ['--port', `${options.port}`], {
      stdio: 'pipe',
      env: {
        DATABASE_DIR: options.database,
        DOC_DIR: resolveBetaFile('docs'),
        WEB_RESOURCES: resolveRoot('web'),
        SIMPLE: resolveRoot('release/simple'),
      },
    });
    cp.stderr!.on('data', (data) => {
      // console.log(data.toString());
    });
    cp.stdout!.on('data', (data: string) => {
      // console.log('data', data.toString());
      if (data.toString().includes('Hamsterbase has started at')) {
        resolve({
          endpoint: `http://localhost:${options.port}`,
          token,
          dispose: async () => {
            cp.kill();
            await fs.rm(options.database, { recursive: true });
          },
        });
      }
    });
  });
}
