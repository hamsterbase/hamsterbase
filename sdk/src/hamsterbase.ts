import { Client, ClientOptions, HamsterBaseRequestLibOption } from './client';
import { Views } from './views/index';
import { WebPages } from './webpages/webpages';

export class HamsterBase {
  static fetchToRequestLib = (fetch: any) => {
    return async (url: string, options: HamsterBaseRequestLibOption) => {
      const response = await fetch(url, {
        method: options.method,
        headers: options.headers,
        body: options.body,
      });
      const status = response.status;
      const body = await response.text();
      return { status, body };
    };
  };

  private client: Client;
  public webpages: WebPages;
  public webpagesView: Views;

  constructor(options: ClientOptions) {
    this.client = new Client(options);
    this.webpages = new WebPages(this.client);
    this.webpagesView = new Views(this.client);
  }
}
