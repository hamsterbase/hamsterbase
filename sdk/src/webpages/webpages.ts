import { Client } from '../client';
import {
  PatchWebPageRequest,
  UploadWebpageRequest,
  UploadWebpageResponse,
  Webpage,
  WebpagesFilterOptions,
  WebpagesList,
  WebpagesListOptions,
  WebpagesSearchOptions,
} from './types';

export class WebPages {
  constructor(private client: Client) {}
  create(data: UploadWebpageRequest): Promise<Webpage> {
    return this.client.post<Webpage>('/webpages', data);
  }

  delete(id: string): Promise<Webpage> {
    if (!id || typeof id !== 'string') {
      throw new Error('invalid webpage id');
    }
    return this.client.delete<Webpage>(`/webpages/${id}`);
  }

  get(id: string): Promise<Webpage> {
    return this.client.get<Webpage>(`/webpages/${id}`);
  }

  getContent(id: string): Promise<string> {
    return this.client.get<string>(`/webpages/${id}/content`);
  }

  update(id: string, newData: PatchWebPageRequest): Promise<Webpage> {
    return this.client.patch<Webpage>(`/webpages/${id}`, newData);
  }

  list(options?: WebpagesListOptions): Promise<WebpagesList> {
    if (options) {
      const searchParams = this.buildFilterOptions(options);
      if (typeof options.sort === 'string') {
        searchParams.append('sort', String(options.sort));
      }
      return this.client.get<WebpagesList>(`/webpages?${searchParams.toString()}`);
    }
    return this.client.get<WebpagesList>(`/webpages`);
  }

  search(options: WebpagesSearchOptions): Promise<WebpagesList> {
    const searchParams = this.buildFilterOptions(options);
    searchParams.append('q', String(options.q));
    return this.client.get<WebpagesList>(`/search/webpages?${searchParams.toString()}`);
  }

  private buildFilterOptions(options: WebpagesFilterOptions) {
    const searchParams = new URLSearchParams({});
    if (typeof options.liked === 'boolean') {
      searchParams.append('liked', String(options.liked));
    }
    if (typeof options.read === 'boolean') {
      searchParams.append('read', String(options.read));
    }

    if (typeof options.host === 'string') {
      searchParams.append('host', String(options.host));
    } else if (Array.isArray(options.host)) {
      options.host.forEach((e) => {
        searchParams.append('host', String(e));
      });
    }

    if (typeof options.labels === 'string') {
      searchParams.append('labels', String(options.labels));
    } else if (Array.isArray(options.labels)) {
      options.labels.forEach((e) => {
        searchParams.append('labels', String(e));
      });
    }
    if (typeof options.ext === 'string') {
      searchParams.append('ext', String(options.ext));
    } else if (Array.isArray(options.ext)) {
      options.ext.forEach((e) => {
        searchParams.append('ext', String(e));
      });
    }
    if (typeof options.page === 'number') {
      searchParams.append('page', String(options.page));
    }
    if (typeof options.per_page === 'number') {
      searchParams.append('per_page', String(options.per_page));
    }
    if (typeof options.annotated === 'boolean') {
      searchParams.append('annotated', String(options.annotated));
    }
    return searchParams;
  }
}
