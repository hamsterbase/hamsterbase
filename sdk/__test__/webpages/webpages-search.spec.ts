import { join } from 'path';
import { afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest';
import { HamsterBase } from '../../src/hamsterbase';
import { WebpagesListOptions, WebsiteExt } from '../../src/webpages/types';
import { createTestServer } from '../server';
import { FixturesId, Fixtures, getBase64Fixture, getPort, resolveRoot } from '../utils';
require('isomorphic-fetch');

describe('test webpages', () => {
  let hamsterbase: HamsterBase;
  let dispose: () => void;
  let port: number;
  beforeAll(() => {
    port = getPort();
  });

  beforeEach(async () => {
    const server = await createTestServer({
      port,
      database: join(resolveRoot('temp'), `${Math.random()}`),
    });
    hamsterbase = new HamsterBase({
      endpoint: server.endpoint,
      token: server.token,
      requestLib: fetch,
    });
    await hamsterbase.webpages.create({
      content: await getBase64Fixture(Fixtures.HamsterBaseDocument_01_mht),
      ext: WebsiteExt.mhtml,
    });
    await hamsterbase.webpages.create({
      content: await getBase64Fixture(Fixtures.HamsterBaseGithubIssue_02_webarchive),
      ext: WebsiteExt.webarchive,
    });
    await hamsterbase.webpages.create({
      content: await getBase64Fixture(Fixtures.HamsterBaseGithubHome_03_html),
      ext: WebsiteExt.html,
    });
    dispose = server.dispose;
  });
  afterEach(async () => {
    await dispose();
  });

  it('001: should support get all pages', async () => {
    const result = await hamsterbase.webpages.search({ q: 'hamsterbase' });
    expect(result.webpages.map((p) => ({ ...p, firstAddTime: null, createTime: null, content: null }))).toMatchSnapshot('search result');
  });

  async function expectList(idList: string[], options: WebpagesListOptions) {
    const result = await hamsterbase.webpages.search({ q: 'hamsterbase', ...options });
    expect(result.webpages.map((p) => p.id)).toEqual(idList);
  }

  it('002: should support filter by liked status', async () => {
    await hamsterbase.webpages.update(FixturesId.HamsterBaseDocument_01_mht, { liked: true });
    await expectList([FixturesId.HamsterBaseDocument_01_mht], { liked: true });
    await expectList([FixturesId.HamsterBaseGithubIssue_02_webarchive, FixturesId.HamsterBaseGithubHome_03_html], { liked: false });
  });

  it('003: should support filter by read status', async () => {
    await hamsterbase.webpages.update(FixturesId.HamsterBaseDocument_01_mht, { read: true });
    await expectList([FixturesId.HamsterBaseDocument_01_mht], { read: true });
    await expectList([FixturesId.HamsterBaseGithubIssue_02_webarchive, FixturesId.HamsterBaseGithubHome_03_html], { read: false });
  });

  it('004: should support filter by host', async () => {
    await hamsterbase.webpages.update(FixturesId.HamsterBaseDocument_01_mht, { link: 'https://z.com' });
    await hamsterbase.webpages.update(FixturesId.HamsterBaseGithubHome_03_html, { link: 'https://test.z.com' });
    await expectList([FixturesId.HamsterBaseDocument_01_mht], { host: 'z.com' });

    await expectList([FixturesId.HamsterBaseDocument_01_mht, FixturesId.HamsterBaseGithubHome_03_html], { host: ['test.z.com', 'z.com'] });
    await expectList([FixturesId.HamsterBaseGithubIssue_02_webarchive], { host: ['github.com'] });
  });

  it('005: should support filter by ext', async () => {
    await expectList([FixturesId.HamsterBaseDocument_01_mht], { ext: WebsiteExt.mhtml });
    await expectList([FixturesId.HamsterBaseGithubHome_03_html], { ext: WebsiteExt.html });
    await expectList([FixturesId.HamsterBaseGithubIssue_02_webarchive], { ext: WebsiteExt.webarchive });

    await expectList(
      [FixturesId.HamsterBaseDocument_01_mht, FixturesId.HamsterBaseGithubIssue_02_webarchive, FixturesId.HamsterBaseGithubHome_03_html],
      { ext: [WebsiteExt.webarchive, WebsiteExt.html, WebsiteExt.mhtml] }
    );
  });

  it('007: should support per_page and page', async () => {
    await expectList(
      [FixturesId.HamsterBaseDocument_01_mht, FixturesId.HamsterBaseGithubIssue_02_webarchive, FixturesId.HamsterBaseGithubHome_03_html],
      { page: 1 }
    );
    await expectList([FixturesId.HamsterBaseDocument_01_mht], { per_page: 1 });
    await expectList([FixturesId.HamsterBaseGithubHome_03_html], { per_page: 1, page: 3 });

    await expectList([FixturesId.HamsterBaseDocument_01_mht, FixturesId.HamsterBaseGithubIssue_02_webarchive], { per_page: 2 });

    await expectList([FixturesId.HamsterBaseGithubHome_03_html], { per_page: 2, page: 2 });
  });
});
