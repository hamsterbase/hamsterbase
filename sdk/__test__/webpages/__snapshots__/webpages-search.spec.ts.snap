// Vitest Snapshot v1

exports[`test webpages > 001: should support get all pages > search result 1`] = `
[
  {
    "annotateCount": 0,
    "content": null,
    "createTime": null,
    "ext": "mhtml",
    "firstAddTime": null,
    "id": "bcf1e35729685a87ce18733080eaf0f80fec0c81a5a4608ef5b3f0272a37851f",
    "labels": "[]",
    "liked": false,
    "link": "https://hamsterbase.com/docs/what-is-hamsterbase.html",
    "read": false,
    "title": [
      {
        "content": "What is ",
        "type": "text",
      },
      {
        "content": "HamsterBase",
        "type": "mark",
      },
      {
        "content": " | ",
        "type": "text",
      },
      {
        "content": "HamsterBase",
        "type": "mark",
      },
      {
        "content": "",
        "type": "text",
      },
    ],
  },
  {
    "annotateCount": 0,
    "content": null,
    "createTime": null,
    "ext": "webarchive",
    "firstAddTime": null,
    "id": "b03e82c2b622df1f039ecba4910f464becf5600b12425ea3d1f65d4c17d70d8d",
    "labels": "[]",
    "liked": false,
    "link": "https://github.com/hamsterbase/hamsterbase/issues",
    "read": false,
    "title": [
      {
        "content": "Issues · ",
        "type": "text",
      },
      {
        "content": "hamsterbase",
        "type": "mark",
      },
      {
        "content": "/",
        "type": "text",
      },
      {
        "content": "hamsterbase",
        "type": "mark",
      },
      {
        "content": " · GitHub",
        "type": "text",
      },
    ],
  },
  {
    "annotateCount": 0,
    "content": null,
    "createTime": null,
    "ext": "html",
    "firstAddTime": null,
    "id": "d4a1267d1f8a8eeb20d7b4b86a9d91839809e083ceb42366847ed5ed3a6cf254",
    "labels": "[]",
    "liked": false,
    "link": "https://github.com/hamsterbase/hamsterbase",
    "read": false,
    "title": [
      {
        "content": "GitHub - ",
        "type": "text",
      },
      {
        "content": "hamsterbase",
        "type": "mark",
      },
      {
        "content": "/",
        "type": "text",
      },
      {
        "content": "hamsterbase",
        "type": "mark",
      },
      {
        "content": ": self-hosted, local-first web archive application.",
        "type": "text",
      },
    ],
  },
]
`;
